### PR TITLE
Add bid per artifact

### PR DESCRIPTION
**Pull request by Rizato**
Support the bid per artifact changes in polyswarmd and contractor. 

TODO
- [x] Update tests
- [x] Modify bid so that bid % sum(bid_portions) is always 0, and no extra 1/10^18ths NCT are given to the arbiters, regardless of verdicts

### CODE REVIEW

File: src/ambassador/filesystem.py [hunk 1]
```diff
@@ -55,20 +55,28 @@ def __init__(self, client, testing=0, chains=None, watchdog=0, submission_rate=3
 
         while True:
             try:
-                filename = random.choice(self.artifacts)
+                num_artifacts = min(255, len(self.artifacts))
```
Comment (mrtizmoatwork) [inline src/ambassador/filesystem.py]: is it supposed to be 255 or 256?

File: src/polyswarmclient/bountiesclient.py [hunk 2]
```diff
@@ -328,7 +328,7 @@ def __init__(self, client):
         Args:
             bounty_guid (str): The bounty which we have asserted on
             index (int): The index of the assertion to reveal
-            nonce (str): Secret nonce used to reveal assertion
+            nonce (str): Secret nonce used to reveal assertio
```
Comment (mrtizmoatwork) [inline src/polyswarmclient/bountiesclient.py]: assertion

File: src/ambassador/filesystem.py (hunk 1 above)
Comment (Rizato) [reply src/ambassador/filesystem.py]: 255, but it should also be len -1. 
num_artifacts is also a bad name, there.

Comment (Rizato) [reply src/ambassador/filesystem.py]: Should be 256. For some reason I was thinking the last value was inclusive. (But the len would still be wrong at that point)

File: src/polyswarmclient/__init__.py [hunk 3]
```diff
@@ -360,6 +360,15 @@ def to_wei(amount, unit='ether'):
     def from_wei(amount, unit='ether'):
         return w3.fromWei(amount, unit)
 
+    @staticmethod
+    def get_artifact_bid_at_(mask, bid, index):
+        if not mask[index]:
+            return 0
+
+        bid_index = len([m for m in mask[:index] if m])
+        logger.critical(mask[:index])
```
Comment (sbnetops) [inline src/polyswarmclient/__init__.py]: Is there a reason to have a critical log here? Everything seems to keep working after this.

File: src/polyswarmclient/abstractmicroengine.py [hunk 4]
```diff
@@ -16,6 +16,10 @@
 logger = logging.getLogger(__name__)
 
 
+class InvalidBidError(Exception):
```
Comment (sbnetops) [inline src/polyswarmclient/abstractmicroengine.py]: Would be great to have a base exception such as `PolyswarmClientException` that all exceptions inherit from. Also, putting all exceptions together in one module would help developers check what kind of error they can expect.

File: src/polyswarmclient/__init__.py (hunk 3 above)
Comment (Rizato) [reply src/polyswarmclient/__init__.py]: Forgot about it. 
Also, need to change the bid_index logic there.

File: src/polyswarmclient/abstractmicroengine.py [hunk 5]
```diff
@@ -103,14 +104,15 @@ def connect(cls, polyswarmd_addr, keyfile, password, api_key=None, testing=0, in
             chain (str): Chain we are operating on
 
         Returns:
-            int: Amount of NCT to bid in base NCT units (10 ^ -18)
+            Bid: Amount of NCT to bid in base NCT units (10 ^ -18)
```
Comment (supernothing) [inline src/polyswarmclient/abstractmicroengine.py]: This should be `list[int]`, shouldn't it?

File: src/polyswarmclient/filters/bountyfilter.py [hunk 6]
```diff
@@ -67,14 +67,14 @@ def is_allowed(self, metadata):
...
-                                                                                     "accept": self.accept}})
+            logger.debug('Metadata not accepted. Skipping artifact', extra={"extra": {"metadata": metadata,
+                                                                            "accept": self.accept}})
             return False
 
         rejected = any([f.filter(metadata) for f in self.reject])
         if self.reject and rejected:
-            logger.info('Metadata rejected. Skipping artifact', extra={"extra": {"metadata": metadata,
-                                                                                 "reject": self.reject}})
+            logger.debug('Metadata rejected. Skipping artifact', extra={"extra": {"metadata": metadata,
```
Comment (supernothing) [inline src/polyswarmclient/filters/bountyfilter.py]: I'm gonna channel some @mrtizmoatwork here and say "single quotes"

File: src/polyswarmclient/filters/bountyfilter.py [hunk 7]
```diff
@@ -67,14 +67,14 @@ def is_allowed(self, metadata):
         accepted = any([f.filter(metadata) for f in self.accept])
 
         if self.accept and not accepted:
-            logger.info('Metadata not accepted. Skipping artifact', extra={"extra": {"metadata": metadata,
-                                                                                     "accept": self.accept}})
+            logger.debug('Metadata not accepted. Skipping artifact', extra={"extra": {"metadata": metadata,
```
Comment (supernothing) [inline src/polyswarmclient/filters/bountyfilter.py]: quotes

**Comment by Rizato** [comment]
@supernothing Can you take another look?

**Pull request merged**